ocaml/quicktest/quicktest_vdi_copy_data_integrity.ml

```diff
@@ -0,0 +1,181 @@
+
+(* ---------------- *
+    Helper functions
+ * ---------------- *)
+
+let with_test test f =
+  Quicktest_common.start test;
+  try
+    f ();
+    Quicktest_common.success test
+  with e ->
+    Quicktest_common.failed test (Printexc.to_string e);
+    raise e
+
+let finally f cleanup =
+  let result =
+    try
+      f ()
+    with e -> begin
+        cleanup ();
+        raise e
+      end
+  in
+  cleanup ();
+  result
+
+let get_domain_zero rpc session_id =
+  Xapi_inventory.inventory_filename := "/etc/xensource-inventory";
+  let uuid = Xapi_inventory.lookup Xapi_inventory._control_domain_uuid in
+  Client.Client.VM.get_by_uuid ~rpc ~session_id ~uuid
+
+let with_attached_vdi rpc session_id vdi mode f =
+  let dom0 = get_domain_zero rpc session_id in
+  let vbd =
+    Client.Client.VBD.create ~rpc ~session_id
+      ~vM:dom0
+      ~empty:false
+      ~vDI:vdi
+      ~userdevice:"autodetect"
+      ~bootable:false
+      ~mode
+      ~_type:`Disk
+      ~unpluggable:true
+      ~qos_algorithm_type:""
+      ~qos_algorithm_params:[]
+      ~other_config:[]
+  in
+  finally
+    (fun () ->
+       Client.Client.VBD.plug ~rpc ~session_id ~self:vbd;
+       finally
+         (fun () -> f ("/dev/" ^ (Client.Client.VBD.get_device ~rpc ~session_id ~self:vbd)))
+         (fun () ->
+            Client.Client.VBD.unplug ~rpc ~session_id ~self:vbd;
+         )
+    )
+    (fun () -> Client.Client.VBD.destroy ~rpc ~session_id ~self:vbd)
+
+let with_open_vdi rpc session_id vdi mode f =
+  with_attached_vdi rpc session_id vdi mode
+    (fun path ->
+       let mode' = match mode with
+         | `RO -> [ Unix.O_RDONLY ]
+         | `RW -> [ Unix.O_RDWR ] in
+       let fd = Unix.openfile path mode' 0 in
+       finally
+         (fun () -> f fd)
+         (fun () -> Unix.close fd)
+    )
+
+(* ---------------- *
+   Integrity tests
+ * ---------------- *)
+
+let write_random_data rpc session_id vdi =
+  let size = Client.Client.VDI.get_virtual_size ~rpc ~session_id ~self:vdi in
+  with_open_vdi rpc session_id vdi `RW
+    (fun fd ->
+
+       let max_writes = 100 in
+       let max_write_length = 2048 in
+
+       let write_random_char_block offset length =
+         let buf =
+           let c = Random.int 256 |> Char.chr in
+           Bytes.make length c
```

Comment:
Same comment as below - ideally this would be more random than just using the same char.